tests/pipeline/test_python_component.py

```diff
@@ -90,10 +93,14 @@ def load(self) -> dd.DataFrame:
             "produces": {"additionalProperties": True},
         },
         "consumes": {},
-        "produces": {"x": {"type": "int32"}, "y": {"type": "int32"}},
+        "produces": {
+            "x": {"type": "int32"},
+            "y": {"type": "int32"},
+            "z": {"type": "int32"},
+        },
     }
 
-    @lightweight_component()
+    @lightweight_component(consumes="generic")
```

Comment:
I think the issue is there are still 3 general options that should be supported without the possibility to mix some of them together.

* consumes is `None` -> **non-generic** component **without specified fields** -> consume all the fields in the dataset schema. Does not have an equivalence in the component spec definition of things. 
* consumes is specified as a `list` -> **non-generic** component **with specified fields** -> start from the fields in the dataset schema and filter based on the fields specified in the list. Equivalent to having specific fields in the component spec. 
* consumes == `generic` -> **generic** component that allows to define **dynamic fields** in the `apply` produces/consumes -> the fields to consumes are the ones specified in the `apply` section. Equivalent to setting `additionalProperties` to `true` in the component spec.

The only solution would be to somehow mix the 1st and 3rd option but this would require us to change the component spec to support both dynamic and specified fields which is not something we currently support

Reply:
@mrchtr updated based on the feedback, consumes is `None` now simply means to consume all fields from previous components. We lose the concept of `additionalProperties` for the `consumes` section in python based components. However I don't think it would be required since we're inferring the fields based on the dataset schema 

This is more evident [here](https://github.com/ml6team/fondant/pull/789/commits/0619c41395dc1963b0a2043683fd1014e8374527#:~:text=consumes%3D%7B%22x%22%3A%20pa.int32()%2C%20%22y%22%3A%20pa.int32()%2C%20%22z%22%3A%20pa.int32()%7D%2C)


```python
    @lightweight_component(
        base_image="python:3.8-slim-buster",
        extra_requires=["pandas", "dask"],
    )
    class CreateData(DaskLoadComponent):
        def load(self) -> dd.DataFrame:
               ....
    dataset = pipeline.read(
        ref=CreateData,
        produces={"x": pa.int32(), "y": pa.int32(), "z": pa.int32()},
    )
    # dataset schema has x,y,z 
        @lightweight_component
    class AddN(PandasTransformComponent):
        def __init__(self, n: int, **kwargs):
            self.n = n

        def transform(self, dataframe: pd.DataFrame) -> pd.DataFrame:
            dataframe["x"] = dataframe["x"].map(lambda x: x + self.n)
            return dataframe

    _ = dataset.apply(
        ref=AddN,
        produces={"x": pa.int32(), "y": pa.int32(), "z": pa.int32()},
        consumes=None,  # This now has to be defined as None since we can't define dynamic fields but we can already infer the schema based on the dataset
        arguments={"n": 1},
    )

I think both options are valuable with small tradeoffs. 
* The previous one would not require any changes between the `consumes` for the apply function for `consumes` but slightly more complex interface for the lightweight component 
* This approach offers more flexibility in the lightweight component with more flexibility, but slightly different way of defining consumes compared to resuable components 

Happy to hear other takes on this @RobbeSneyders @GeorgesLorre

Reply:
Agree with not supporting `consumes=generic`. The `additionalProperties: true` is only needed because the user cannot easily change the schema of a reusable component. But they can easily change the schema of a lightweight one. So there's no need for it here.

Lightweight Python components can still be implemented in a generic way without it. It just means that the implementation of the component depends on the `consumes` argument it receives.